### ISSUE TITLE
Cannot set events with a user preferred time zone

### ISSUE THREAD

**Issue by stevepiercy** [plone/volto#5324]
Cross-posting from https://github.com/plone/plone.org/issues/141

It is not possible to create local events with a local time zone in plone.org or in the [Volto demo](https://demo.plone.org/), but I can in the [Classic UI demo](https://classic.demo.plone.org/en).

## Classic UI Demo behavior

- Login to https://classic.demo.plone.org/en as `admin`/`admin`.
- Go to [Date and Time control panel](https://classic.demo.plone.org/@@dateandtime-controlpanel).
- Observe the only available time zone is Europe/Berlin by default.
- Add any time zone, such as `America/Los_Angeles`, and click Save.
- Click Home to return home.
- Create a new event. The event inherits the site's default time zone.
- Go to User Preferences, select the available time zone, `America/Los_Angeles`, and click Save.

<img width="734" alt="Screenshot 2023-10-15 at 7 05 03 PM" src="https://github.com/plone/plone.org/assets/102112/0f657cd3-c080-4252-ac0b-fec3792b91f4">

<img width="696" alt="Screenshot 2023-10-15 at 7 04 58 PM" src="https://github.com/plone/plone.org/assets/102112/3738df95-996d-4f01-ba8a-3c12de556d06">

## Plone.org and Volto Demo behavior

- Repeat all the steps above. Observe that the User cannot set their preferences for a time zone, and must therefore rely upon the site preference for a time zone when creating events.

I don't know whether this is a bug in Volto or some other dependency.

**Comment by ashvaneetk** [comment]
Assign me this issue

**Comment by nileshgulia1** [comment]
@ashvaneetk please read and follow [Contributing to Plone](https://6.docs.plone.org/contributing/index.html) and [First-time contributors](https://6.docs.plone.org/contributing/first-time.html).

**Comment by tiberiuichim** [comment]
@stevepiercy AFAIK, in Volto the datetime is TZ-aware, and it's extracted from the user's browser.

**Comment by stevepiercy** [comment]
We want to be able to schedule an event in a different time zone than the user's browser, allowing the user to set a preferred time zone. The use cases include traveling outside one's local time zone or scheduling an event that takes place outside one's local time zone. The Volto implement does not allow either use case, whereas Classic UI does allow this. In America this is a serious problem, for example, where an organization is headquartered in the Eastern time zone but wants to schedule an event in the Pacific time zone.

**Comment by IshaanDasgupta** [comment]
@stevepiercy is adding a timezone options in the user preferences okay ?
![Screenshot from 2023-11-24 15-04-01](https://github.com/plone/volto/assets/62870756/9de651f0-b01d-4218-a88f-5a0fbd0b29b5)

**Comment by stevepiercy** [comment]
@IshaanDasgupta Volto should have feature parity with Classic UI.

**Comment by IshaanDasgupta** [comment]
@stevepiercy sorry but if I am understanding the problem wrong let me know.

**Comment by stevepiercy** [comment]
That is half the problem. Once the user sets a preferred timezone, we need to ensure that events that the user creates are in that same preferred time zone.

One more thing that was pointed out to me by @ericof is that Volto does not provide an .ical link for events. That should also be included. But maybe that should be a separate issue? We definitely need that in Volto, especially for PloneConf and sprints.

**Comment by IshaanDasgupta** [comment]
Yes, I am planning to overwrite the timezone(by getting it from the redux store) whenever the user selects a time also I am planning to add a text below the date-time picker that the time is being selected in the respective timezone. If no preference is selected then that text will not be displayed and the default timezone will be used.
Also, I am not familiar with .ical links so, I'll create a new issue for that and after implementing this, I'll look into that if no one has until then.

Is that alright ?

**Comment by stevepiercy** [comment]
I don't know whether getting event information from the redux store is the best way to get something that is stored in the backend. Although as @tiberiuichim points out about Volto extracting the timezone from the user's current location, I think that method prevents an editor from creating events in other time zones. I would think the best way would be to use the [Plone REST API event endpoint](https://6.docs.plone.org/plone.restapi/docs/source/endpoints/content-types.html#event). It would be worth identifying how Classic UI handles events as well.

**Comment by ericof** [comment]
> That is half the problem. Once the user sets a preferred timezone, we need to ensure that events that the user creates are in that same preferred time zone.
> 
> One more thing that was pointed out to me by @ericof is that Volto does not provide an .ical link for events. That should also be included. But maybe that should be a separate issue? We definitely need that in Volto, especially for PloneConf and sprints.

I corrected myself, Volto does provide the .ics download link (as can be seen in this image "Baixar evento"
![Screenshot 2023-11-24 at 09-48-51 Plone Symposium South America 2023](https://github.com/plone/volto/assets/306014/85110e07-7a30-4071-a918-16a9c3d210dc) )

**Comment by IshaanDasgupta** [comment]
@stevepiercy as you mentioned, I went through the classic UI once again. In the classic 
UI the event date-time is always displayed in the timezone in which it was created (so if my preference is Europe/Berlin and the event was created in America/Los_Angeles the date-time in the event would be displayed in the America/Los_Angeles and not be converted to my preferred timezone). Do we want the same in volto ? 
Or we could keep all date-times in UTC and when rendering we could convert them to the default/preferred timezone (personally this makes more sense but then it would not be similar to the classic UI).

Also I was referring to getting the preferred timezone from the redux store not any event information.

**Comment by stevepiercy** [comment]
> if my preference is Europe/Berlin and the event was created in America/Los_Angeles the date-time in the event would be displayed in the America/Los_Angeles and not be converted to my preferred timezone

This is exactly what we want. In most use cases, the event's time zone will be that of its location. One exception is where there is no physical location, such as online only events such as Plone Tune Ups. In that use case, it would probably be best for the admin to enable UTC and editors set the event start and end dates and times in UTC.

As an aside, did I just discover something that has been missing in Plone for all time? I think that relying on the user's preferred or site's default time zone setting is wrong, but this issue will work with what exists now in Plone. IMO both Classic UI and Volto should allow an editor to set the event's time zone and location, regardless of their preferred time zone setting, instead of inheriting it from the user's or site's time zone setting. Thus I could create an event in the Eastern US time zone while residing in Pacific US time zone. The event would be displayed in its time zone, not the user's or site's.

**Comment by IshaanDasgupta** [comment]
@stevepiercy sorry for the late reply, I was a bit caught up.
Yes, I totally agree that an editor should be able to create events in timezone regardless of his preference. So, to implement that, I could add a timezone drop down when creating the event so the editor is able to select any timezone for that event but, then what would be the meaning of having preference ?
We could make the preference the default timezone for the user for example I am in India but if I were to chose a US timezone, all things will be converted to the US time (except for the date-time which have a specific timezone associated with them) or we could skip implementing the user preference timezone for now.

Is the drop-down idea okay or you have any other ideas for the implementation or would you recommend to implement it in a  similar way to Classic UI ?

**Comment by thet** [comment]
@stevepiercy when plone.app.event was created, the plan was to support event-related timezones. There would even the possibility to support different time zones for the start and end date.

It wasn't implemented because lack of a design proposal and lack of time.

**Comment by davisagli** [comment]
@stevepiercy https://www.youtube.com/watch?v=XZlPXLsSU2U is a good fairly recent talk on handling timezones in Python, including the new `zoneinfo` module

**Comment by thet** [comment]
@stevepiercy that's not bad news that pytz is obsolete. It's a good library but also importing and loading it took several seconds as the database behind pytz is huge. OK, maybe importing tzdata in core Python is still slow, I‌ don't know. However, having fewer dependencies is always good.

@stevepiercy and having you as a champion on implementing event-related timezone support is definitely good news. I‌ can probably assist if there are any questions. I was back then the main contributor on plone.app.event.

Regarding Volto: I don't think they need this feature implemented in ClassicUI before they can implement it. The foundations are already here. The start and end dates are stored with timezone information (thus, they could also be different). For the catalog everything is translated to UTC. It's mainly an UI issue.

IIRC there is a problem with JavaScript and timezones in general, that core JavaScript has poor timezone support. Date/time only know UTC offsets. But that is not enough to calculate daylight saving time changes, for example, which is regulated by country laws.
Volto therefore needs to separately serialize the timezone information for the REST API endpoints dealing with datetimes. We discussed this issue 5 years ago in Bonn. Something was done, also by myself (cant remember what OTTOMH) but apparently there is still not timezone information in the event's REST API endpoint, looking at the resonse on https://demo.plone.org/test-event

**Comment by IshaanDasgupta** [comment]
@stevepiercy if its okay, I would like to work on the implementation part in volto.

**Comment by rnunez80** [comment]
The issue with the time in volto is more than just the events and recurrences, it also affects effective and expiration, and also volto-form-block

**Comment by stevepiercy** [comment]
@IshaanDasgupta thank you for stepping up. As @thet and I both mentioned, you will need to enhance the [event REST API endpoint](https://6.docs.plone.org/plone.restapi/docs/source/endpoints/content-types.html#event), as well as work on the Volto frontend part. Please let me know if you need any direction for how to proceed.

@rnunez80 can you elaborate what you mean? I think you mean the publishing datetimes for "effective" and "expiration" for all content types, but I am not sure. I also don't know what is volto-form-block.

**Comment by thet** [comment]
That's right @rnunez80 and @IshaanDasgupta very nice if you want to step up for this, but since this is not a simple bugfix but needs some planning and specification you need to coordinate with the other Volto people.

**Comment by IshaanDasgupta** [comment]
@stevepiercy and @thet thanks a lot, right now I am looking into the rest APIs as I not familiar with them

> I'd like to work on this, at least creating a design proposal or PLIP and being its champion, if not working the code.

The PLIP would help a lot, I know it's quite a large issue and I am totally fine working alongside someone but the issue looks very interesting and I would love to contribute.

**Comment by rnunez80** [comment]
> @rnunez80 can you elaborate what you mean? I think you mean the publishing datetimes for "effective" and "expiration" for all content types, but I am not sure. I also don't know what is volto-form-block.

[Here is a video of the effective date issue](https://youtu.be/EqkZz9AEDRw)
in this video I set the effective date about 27 min before the current time but it shows as effective and will not show on the site for the next ~4 hours due to time zone (utc -4)

[Date widget Issue in volto-form-block](https://youtu.be/ZDP952MM3DQ)
date selected 2023-12-04, date on email 2023-12-03 (I believe is because the date selected would be 2023-12-04 12am but due to time zone -4 it is changed to 2023-12-03 8pm)

**Comment by davisagli** [comment]
@rnunez80 The time zone issue with effective and expiration date is a separate issue, not related to this one which is about storing a time zone as part of the Event content type. There is a PR in progress to fix it: https://github.com/plone/plone.restapi/pull/1192

The issue with volto-form-block should be submitted in that repository.

**Comment by IshaanDasgupta** [comment]
was looking into the code of plone.restapi and finally found the problem, due to the conversion of date to ISO format date, the tz info was lost.
@thet found [this commit](https://github.com/plone/plone.restapi/commit/d0df8512a720e4bf42468c6245210ab03d4d5ba5#diff-bd2bdc9cbb6e96b0cf390ceed217065fca13c777e016414bcfdd5a3e394408a5) by you, the problem is in the 'datetimelike_to_iso' function, there is also a comment by @gotcha as to why did you keep the tz code commented out, could you let me know the reason for that. Like will there be any issues if we were to include the tz info which you noticed at the time ?

Apart from this, I wanted to ask a few things : 

- There are 2 things which we can do to solve this, first is to just make the rest endpoint return dates with tz info but if we were to do that, we would not be able to create event dates which don't need a timezone like online events as @stevepiercy mentioned earlier. If we want that to be possible, is there any other solution apart from adding a separate timezone attribute to the event content type ?
- If we were to change the plone.restapi to return datetime with tz info would that cause problems to existing volto or any custom frontends which use plone.restapi ? (I tried to look into this and did not find any problems as the datetime always had the tzinfo internally it was just not returned in the response)

**Comment by stevepiercy** [comment]
> * we would not be able to create event dates which don't need a timezone like online events as @stevepiercy mentioned earlier.

That is not correct. All events content types should have timezones, using the fallback scheme, as I mentioned in my [earlier comment](https://github.com/plone/volto/issues/5324#issuecomment-1832565076).

> If we were to change the plone.restapi to return datetime with tz info would that cause problems to existing volto or any custom frontends which use plone.restapi ?

Good question. I assume it would be tested in Volto, `plone.restapi`, and `plone.app.event`. I'll include that in the PLIP I am drafting as a potential risk.

**Comment by thet** [comment]
I just got reminded that we already support per-field timezones. It's done like that:

```
from plone.app.event.base import default_timezone
from plone.app.z3cform.widget import DatetimeFieldWidget
from plone.autoform import directives
from zope import schema
from plone.supermodel import model

class IExampleDate(model.Schema):
    directives.widget("example_date", DatetimeFieldWidget, default_timezone=default_timezone)
    example_date = schema.Datetime(
        title="Example date",
        required=False,
    )
```
This renders something like:

![image](https://github.com/plone/volto/assets/170891/70c830ac-c9ac-463c-beac-44b20b291977)

Only, it seems like full of bugs. I'm currently evaluating (but not yet fixing) the situation.
